### ISSUE TITLE
[Behaviour Change] Generalisation error metric is mislabelled

### ISSUE THREAD

**Issue by rpreen** [AI-SDC/AI-SDC#468]
`StructuralAttack` reports the "generalisation error" but it is actually reporting the accuracy gap between the train and test set (it is also inverted so returns a negative value). Generalisation error is conventionally understood to be the test set error.

**Comment by shamykyzer** [comment]
hello @rpreen , would it be okay if I work on this issue?

**Comment by rpreen** [comment]
Sure; I think we want the "generalisation gap" which is the test minus train error.

Slightly complicating things (and maybe some of this should be deferred for #414 but it seems simple enough) is that to also support regression models we will need to dynamically adapt the calculation depending on if it's a regressor or classifier (classification error could just be `1 - score()` when `score()` returns the accuracy; and I think the `PytorchModel` `score()` would need to be updated to return the MSE if it's a regressor. We can probably determine the model type on the fly: for scikit-learn, we can just check the `_estimator_type` attribute, and for PyTorch, we could infer it based on the output data types (int vs float).

**Comment by rpreen** [comment]
On further thought, I think it makes sense to keep this one simple and just address fixing it for classification models.

Dealing with the calculation for regression models should be part of adding regression model support since setting the model type should probably just be done once and tracked as an attribute, probably in the `Model` abstract class.
